### DISCUSSION TITLE
Add entryTypes to Graphql root

### DISCUSSION THREAD

**Discussion by gopeter** [craftcms/cms#9959] [Ideas]
We're using Craft CMS as a headless CMS for a Next.js frontend. We're also using [`graphql-coden`](https://www.graphql-code-generator.com/) to create typed hooks for our Graphql queries. But since `entries` could be any entryType, a query like 

```graphql
query {
  entries {
    ... on homepage_homepage_Entry {
      id
    }
  }
}
```

would result in `entries: Array<{ id: string } : {}>`, because we have a lot of other entry types:

![image](https://user-images.githubusercontent.com/1242960/137090202-3965b998-4d42-47e8-a412-533f002102df.png)

For sure we can use a section filter to get only results with the `homepage` entryType, but the codegen does not know that this filter is going to return one (or more) specific type(s) only. 

Now writing a typed frontend becomes a mess, since we have to double check the typenames everywhere and extract the right types from huge unions. It would be nice if we could have something like:

```graphql
query {
  typedEntries {
    homepage {
      id
    }
    faqPages {
      id
    }
  }
}
```

`typedEntries` may not be the best name, but I think you know what I mean 😄 Is this something that you may can think of? Or should I add such root nodes by myself?

**Comment by andris-sevcenko** [comment]
I understand where this is coming from, for sure. My current worry about this is that some sites would get hundreds of extra queries. Maybe the solution would be to hide queries that return a single type behind a configuration option? I'll bring this up internally and we'll talk about it. I should get back to you by the end of the week about this.

**Reply by gopeter** [reply]
Hi @andris-sevcenko. Wow, I'm always impressed by how fast you're responding to issues and discussions! 😄
I think it would suffice it we could configure this, even if it's just an opt-in array for some types in a configuration file like `config/general.php`.

**Comment by gopeter** [comment]
Hi @andris-sevcenko. There's another point I want to add and which relates to my initial query.

I've created a field where I can choose an entry, but with just a single entry type:

![image](https://user-images.githubusercontent.com/1242960/137288036-427dc861-85d6-4169-a3a4-b8e4f22c3e30.png)

But if I want to query this with Graphql, the Schema does not know that this can be only one type:

![Bildschirmfoto 2021-10-14 um 11 16 06](https://user-images.githubusercontent.com/1242960/137288516-6d7257be-a7f1-4c62-bd28-aeae7fa2ea4f.png)

It would be great if the only possible entry type for `destination` would be `destinations_destinations_Entry`, since I did not selected any other entry types for this relation.

**Reply by andris-sevcenko** [reply]
This is easy to implement but tricky in its own way. Consider a site with a public GraphQL API. Whether it's for public consumption or for a Vue SPA, who knows. Let's just say there's a public GraphQL API. 

Now, in Craft there's a `destination` field that allows only entries from the Destination section that has only a `destination` entry type. Craft goes through with the effort (more on that later) and generates the GQL Schema so that the `destination` field is of type `destination_destination_Entry` type, not `EntryInterface`.

The site grows and, one day content editors figure that, for marketing purposes, they need "experience" as another destination entry type. Every field is the same, except there is not, let's say, `location` field on the `experience` entry type.

At this point, no matter how Craft compensates and whether it reverts back to the `EntryInterface` type or decides to make a Union of the two types, any query that queries for the `location` field and does not use the `.. destination_destination_Entry` safeguard (which was not needed until this change), will break.

So, to conclude, it's possible to implement it, but the opportunity cost is losing the flexibility of changing the content model on the backend with little consequences.

Another drawback is a small overhead of CPU when figuring out whether EntryInterface should be used or something else. It's tiny, for sure, but on large sites it will amount to quite a bit and sites with complex content models are already are a bit in narrow spaces when it comes to [generating the schema](https://github.com/craftcms/cms/discussions/8968).

**Reply by gopeter** [reply]
Hmmmm, I understand, but what's about keeping the unions? In this case, the `destination` field could be just `destination_destination_Entry` or `destination_experience_Entry`, but not `search_search_Entry`, since I've selected just "Destinations" as possible source.

**Reply by andris-sevcenko** [reply]
A union could be used. It, however, would still require you to use the `... on specific_type`, so you're not gaining much in the sense of cleaner queries. You are getting more hints about the type of content, though, which is maybe useful by itself. Would union work better for you?

Also, to be clear, if/when we add this, it's likely that it'll be for Craft 4.0, not 3.7.x, since it's a bit of a structural change and feels more right for 4.0.

**Reply by gopeter** [reply]
That would be fine, since it would be the same source at least. If `destination` and `experience` share a lot of fields, Typescript would not complain about a missing field which does not exists on `search_search_Entry` for example. 

This in combination with real single root entry types (my initial request) would make working with the Graphql API much easier! 

At the moment I'm thinking about dropping TS at all in our current project, because the frontend is full of ... 

```tsx
 const destination = destinationOfTheMonth.destination[0] as Extract<
          NonNullable<NonNullable<NonNullable<NonNullable<DestinationsOfTheMonthProps['destinationsOfTheMonth']>[0]>['destination']>[0]>,
          { __typename?: 'destinations_destinations_Entry' }
        >
```

... which gives us at least some kind of autocomplete, but leads the purpose of Typescript ad absurdum. And waiting for Craft CMS 4 is not an option 😬

However, the example above suffers from other little issues too, like nullable arrays. I know that the Relay spec also says, that [edges and nodes should be nullable](https://stackoverflow.com/a/57908071/1184904), but I'm not sure if this applies to Craft CMS' implementation also? Let's take this query (it's just a part) as example:

```graphql
children {
  id
  level
  ... on contentBuilder_destinationsOfTheMonth_BlockType {
    destinationsOfTheMonth {
      ... on destinationsOfTheMonth_BlockType {
        id
      }
    }
  }
}
```

I would not expect that `destinationsOfTheMonth` could be null at all. Either it's an empty array or not, but not null. But I think that's a completely different discussion 😄

**Reply by gopeter** [reply]
Oh, sorry, maybe this is a bad example, since this comes from the Super Table plugin. But looking at the whole query, it applies to the built-in entry selection also: 

```graphql
children {
  id
  level
  ... on contentBuilder_destinationsOfTheMonth_BlockType {
    destinationsOfTheMonth {
      ... on destinationsOfTheMonth_BlockType {
        id
        destination {
          id
        }
      }
    }
  }
}
``` 

In this case, I would not expect that `destination` could be `null`, but at most an empty array.

**Reply by andris-sevcenko** [reply]
So are you saying that NonNullable should be dropped in such a case, is that correct?

I hadn't really thought of that, but that does make sense.

**Reply by gopeter** [reply]
Yes. There are many cases where we would expect an empty array and therefore never `null`, but this depends on how Craft CMS is handling errors ([since this is the reason Relay expects that edges/nodes to be nullable](https://stackoverflow.com/questions/57901668/how-to-understand-null-edges-and-nodes-in-graphql/57908071#57908071)).

There are a few more fields that I wouldn't expect to be null at all. By creating a new draft, I always get a value for a lot of fields (just a few examples):

![image](https://user-images.githubusercontent.com/1242960/137445530-443f039d-290f-4129-afe0-18eec8649130.png)

Looking at the generated Type, every single field could be null ... which makes it tedious to work with 😄🙈 For sure there are cases where fields could be null, especially content fields, since we can save drafts without filling out any field. But there are also fields, like `id` or `language`, which shouldn't be nullable in the Graphql schema.

![image](https://user-images.githubusercontent.com/1242960/137445890-db04a30f-3689-45b1-867e-b229c843a239.png)

**Reply by andris-sevcenko** [reply]
It's tricky. Craft fields can have null values, as opposed to, say, an empty string. Also, if we're looking at drafts, then _none_ of the fields have to pass validation yet, so, even if a field is required, I can't really enforce it to be non-null, because drafts don't require that field.

And in that case, it's a whole other discussion whether returning an empty string for a value that was not specified yet by the user is the right approach.

The more flexible the system becomes, the less TypeScript likes it 🤷

**Reply by gopeter** [reply]
Yes, that's what I meant with

> For sure there are cases where fields could be null, especially content fields, since we can save drafts without filling out any field.

But ID, language, etc, shouldn't be null, even for drafts without any content at all. Or do I miss something here? 🤔

**Reply by andris-sevcenko** [reply]
> Or do I miss something here?

No, sorry, I just wasn't careful enough when reading your last message. You are right. I'll see if I can tighten things up a little bit around here!

**Reply by gopeter** [reply]
Thank's a lot 😊

**Reply by andris-sevcenko** [reply]
@gopeter we'll tighten the definitions, but that'll have to happen in Craft 4.0, after all.

Changing around several signatures at once does not feel like a Craft 3.7.x change, as it's a bit too significant to add it to such a minor release. 

Since there's not going to be a Craft 3.8, the only other option left is to add that to Craft 4.0, which is currently slated to be in Beta in Q1 2022.

Meanwhile, you can modify the definitions using the `EVENT_DEFINE_GQL_TYPE_FIELDS` event. I realize it's not very convenient, but it's probably more convenient than me breaking everything and rushing to implement a change across all the API :)

**Reply by andris-sevcenko** [reply]
I [tightened some of the field types](https://github.com/craftcms/cms/commit/50803fd941981c8370ff53dc488e1e3b1484c5bf) for 4.0.

The union thing, however, will have to wait some more time. Unfortunately, it's impossible to return a Union and Interface at the same time (even if the union is a subset of types implementing the interface), which means that every query that lists an interface field without fragments will break.

These changes have been pushed back to Craft 5, since that's when the content model will change a bit which will directly affect GraphQL queries and likely require people to rewrite them. We can't really ask people to do that for two subsequent Craft releases, so we have to push the "unions instead of interfaces" changes back to v5 as well.

I'm still looking into adding additional queries that return unions to Craft 4, though, since that is an additional change, not an in-place change.
